internal/fs/type_cache_test.go

```diff
@@ -0,0 +1,390 @@
+	"path"
+	"time"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/config"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/storageutil"
+
+	"github.com/jacobsa/oglematchers"
+	. "github.com/jacobsa/ogletest"
+)
+
+////////////////////////////////////////////////////////////////////////
+// Common
+////////////////////////////////////////////////////////////////////////
+
+type typeCacheTestCommon struct {
+	fsTest
+}
+
+var (
+	// The following should be configured for different tests
+	// differently inside SetUpTestSuite as these need to
+	// set for mount itself.
+
+	// ttlInSeconds is equivalent of metadata-cache:ttl-secs in config-file.
+	ttlInSeconds int64
+
+	// typeCacheMaxSizeMb is equivalent of metadata-cache:type-cache-max-entries in config-file.
+	typeCacheMaxSizeMb int
+)
+
+func (t *typeCacheTestCommon) SetUpTestSuite() {
+	t.serverCfg.MountConfig = config.NewMountConfig()
+	t.serverCfg.MountConfig.MetadataCacheConfig = config.MetadataCacheConfig{
+		TypeCacheMaxSizeMB: typeCacheMaxSizeMb,
+		TtlInSeconds:       ttlInSeconds,
+	}
+
+	// Call through.
+	t.fsTest.SetUpTestSuite()
+}
+
+////////////////////////////////////////////////////////////////////////
+// Specific test classes
+////////////////////////////////////////////////////////////////////////
+
+type TypeCacheTestWithMaxSize1MB struct {
+	typeCacheTestCommon
+}
+
+func (t *TypeCacheTestWithMaxSize1MB) SetUpTestSuite() {
+	ttlInSeconds = 30
+	typeCacheMaxSizeMb = 1
+
+	t.typeCacheTestCommon.SetUpTestSuite()
+}
+
+type TypeCacheTestWithZeroCapacity struct {
+	typeCacheTestCommon
+}
+
+func (t *TypeCacheTestWithZeroCapacity) SetUpTestSuite() {
+	ttlInSeconds = 30
+	typeCacheMaxSizeMb = 0
+
+	t.typeCacheTestCommon.SetUpTestSuite()
+}
+
+type TypeCacheTestWithZeroTTL struct {
+	typeCacheTestCommon
+}
+
+func (t *TypeCacheTestWithZeroTTL) SetUpTestSuite() {
+	ttlInSeconds = 0
+	typeCacheMaxSizeMb = 1
+
+	t.typeCacheTestCommon.SetUpTestSuite()
+}
+
+type TypeCacheTestWithInfiniteTTL struct {
+	typeCacheTestCommon
+}
+
+func (t *TypeCacheTestWithInfiniteTTL) SetUpTestSuite() {
+	ttlInSeconds = -1
+	typeCacheMaxSizeMb = 1
+
+	t.typeCacheTestCommon.SetUpTestSuite()
+}
+
+func init() {
+	RegisterTestSuite(&TypeCacheTestWithMaxSize1MB{})
+	RegisterTestSuite(&TypeCacheTestWithZeroCapacity{})
+	RegisterTestSuite(&TypeCacheTestWithZeroTTL{})
+	RegisterTestSuite(&TypeCacheTestWithInfiniteTTL{})
+}
+
+// //////////////////////////////////////////////////////////////////////
+// helpers
+// //////////////////////////////////////////////////////////////////////
+func (t *typeCacheTestCommon) testNoInsertion() {
+	const name1 = "foo"
+	const contents = "taco"
+	var fi fs.FileInfo
+	var err error
+
+	// Create a file object in GCS.
+	fileObject, err := storageutil.CreateObject(
+		ctx,
+		bucket,
+		name1,
+		[]byte(contents))
+
+	ExpectEq(nil, err)
+	ExpectNe(nil, fileObject)
+
+	// Stat-call with file object. It should
+	// be stattable, bypassing type-cache, as a file.
```

Comment:
You mean stattable-> statable?

Reply:
There is no dictionary word for this. All we mean here is that we can call `stat` system call on the object. I didn't like use the of the word `statable` as it already has a different [meaning](https://www.google.com/search?q=meaning+of+statable) .

Reply:
changed.